### PR TITLE
Document function rules for File-based Access Control

### PR DESCRIPTION
**Pull request by huberty89**
## Description
Update documentation for https://github.com/trinodb/trino/pull/13713

### CODE REVIEW

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 1]
```diff
@@ -25,6 +25,13 @@ cluster in a single JSON file.
 Configuration
 -------------
 
+.. warning::
+
+    From Trino 401 the default access to PTFs (Polymorphic Table Function) has
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:30]:
```suggestion
    As of Trino 401, the default access to PTFs (Polymorphic Table Function) has
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 2]
```diff
@@ -25,6 +25,13 @@ cluster in a single JSON file.
...
+    changed from always denied to allowed. To mitigate that it's require to add
+    a ``function`` :ref:`rule <system-file-function-rules>` to deny ``TABLE``
+    function kind.
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:33]:
```suggestion
    changed from always denied to allowed. To mitigate this, you should add
    a ``function`` :ref:`rule <system-file-function-rules>` to deny ``TABLE``
    functions.
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 3]
```diff
@@ -106,6 +113,15 @@ DELETE FROM                          all                delete
 UPDATE                               all                update
 ==================================== ========== ======= ==================== ===================================================
 
+Permissions required for executing functions and PTFs (Polymorphic Table Function):
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:116]: If you're adding a new table, please make it a list-table. You can search in the docs for `list-table` for examples and formatting.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 4]
```diff
@@ -117,10 +133,10 @@ items do not need to already exist as any potential permission makes the item
 visible. Specifically:
 
 * ``catalog``: Visible if user is the owner of any nested schema, has
-  permissions on any nested table, or has permissions to set session properties
-  in the catalog.
+  permissions on any nested table or PTF (Polymorphic Table Function) or has
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:136]:
```suggestion
  permissions on any nested table or PTF (Polymorphic Table Function), or has
```

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 2 above)
Comment (colebow) [reply docs/src/main/sphinx/security/file-system-access-control.rst:33]: @m57lyra may have a better idea for how to clearly word this warning.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 5]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
       ]
     }
 
+.. _system-file-function-rules:
+
+Function rules
+^^^^^^^^^^^^^^
+
+These rules control the ability of a user to execute SQL functions, including
+PTFs (Polymorphic Table Functions).
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:356]:
```suggestion
These rules control the user's ability to execute SQL functions.
```
Same thing as above where I don't think we need to say functions _and_ PTFs.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 3 above)
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:116]:
```suggestion
Permissions required for executing functions:
```
I don't think we need to be specifically calling out PTFs - they're also functions, and while they're neat, they don't need super special treatment relative to other functions.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 6]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
...
+* ``schema`` (optional): regex to match against schema name. Defaults to ``.*``.
+* ``function`` (optional): regex to match against function names. Defaults to
+  ``.*``.
+* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
+* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
+  ``TABLES``, ``WINDOW``. When user define a rule for ``AGGREGATE``, ``SCALAR``
+  or ``WINDOW`` function, ``catalog`` and ``schema`` fields are disallowed.
+
+For example, to restore previous behavior (all Polymorphic Table Functions from any
+catalogs is denied) you can use the following rules:
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:374]:
```suggestion
For example, to restore previous behavior (all polymorphic table functions from any
catalog are denied), you can use the following rules:
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 7]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
...
+          "privileges": [],
+          "function_kinds": [
+            "TABLE"
+          ]
+        }
+      ]
+    }
+
+To allow executing ``query`` polymorphic table function from any catalogs by
+``admin`` user.
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:401]:
```suggestion
To allow use of the ``query`` polymorphic table function from any catalog by
an ``admin`` user:
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 8]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
...
+
+.. note::
+
+    ``query`` polymorphic table function should be used by admins in a first
+    place.
```
Comment (colebow) [inline docs/src/main/sphinx/security/file-system-access-control.rst:406]: Do we really need this note? It's a little jarring, but If so...

```suggestion
.. note::

    It is recommended to only allow admins to use the ``query`` polymorphic table
    function.
```

**Review by colebow — CHANGES_REQUESTED** [review]
I've made an initial pass. Defer to @m57lyra if she suggests anything in the same places, as she's better at this than I am.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 9]
```diff
@@ -633,6 +736,22 @@ These rules apply to ``filter_environment`` and ``mask_environment``.
...
+* ``user`` (optional): regex to match against user name. Defaults to ``.*``.
+* ``role`` (optional): regex to match against role names. Defaults to ``.*``.
+* ``group`` (optional): regex to match against group names. Defaults to ``.*``.
+* ``schema`` (optional): regex to match against schema name. Defaults to ``.*``.
+* ``function`` (optional): regex to match against function names. Defaults to
+  ``.*``.
+* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
+* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
+  ``TABLES``, ``WINDOW``. When user define a rule for ``AGGREGATE``, ``SCALAR``
+  or ``WINDOW`` function, ``catalog`` and ``schema`` fields are disallowed.
```
Comment (m57lyra) [inline docs/src/main/sphinx/security/file-system-access-control.rst:753]: This appears to duplicate lines 358 to 371. Is this intentional @huberty89 ? If so, all suggestions and comments there apply here also.

**Review by m57lyra — COMMENTED** [review]
Some things to take care of.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:406]: > Do you mean that only admins should have this privilege?

In short yes. `query` PTF bypass the whole access control and should be used carefully.
To make access more strict, `admin` should create a `SECURITY DEFINER` `views` which execute specific query using `query` PTF so user can use only those defined by `admin`.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 9 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:753]: Everything here is the same expect `role` and `catalog` - I forgot to remove `role` entry before.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 1 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I mix this with https://github.com/trinodb/trino/pull/14774#discussion_r1009694755

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 5 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Mixed with https://github.com/trinodb/trino/pull/14774#discussion_r1009696805

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 3 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 6 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I used the suggestion from https://github.com/trinodb/trino/pull/14774#discussion_r1009704916

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 7 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I used the suggestion from https://github.com/trinodb/trino/pull/14774#discussion_r1009706978

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 8 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I mix this with https://github.com/trinodb/trino/pull/14774#discussion_r1009685115

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 10]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
       ]
     }
 
+.. _system-file-function-rules:
+
+Function rules
+^^^^^^^^^^^^^^
+
+These rules control the ability of a user to execute SQL functions, including
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Mixed with https://github.com/trinodb/trino/pull/14774#discussion_r1009690830

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 11]
```diff
@@ -117,10 +133,10 @@ items do not need to already exist as any potential permission makes the item
 visible. Specifically:
 
 * ``catalog``: Visible if user is the owner of any nested schema, has
-  permissions on any nested table, or has permissions to set session properties
-  in the catalog.
+  permissions on any nested table or PTF (Polymorphic Table Function) or has
+  permissions to set session properties in the catalog.
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Mixed with https://github.com/trinodb/trino/pull/14774#discussion_r1009688718

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 12]
```diff
@@ -117,10 +133,10 @@ items do not need to already exist as any potential permission makes the item
...
 * ``schema``: Visible if the user is the owner of the schema, or has permissions
-  on any nested table.
+  on any nested table or PTF (Polymorphic Table Function).
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 13]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
...
+* ``user`` (optional): regex to match against user name. Defaults to ``.*``.
+* ``role`` (optional): regex to match against role names. Defaults to ``.*``.
+* ``group`` (optional): regex to match against group names. Defaults to ``.*``.
+* ``catalog`` (optional): regex to match against catalog name. Defaults to
+  ``.*``.
+* ``schema`` (optional): regex to match against schema name. Defaults to ``.*``.
+* ``function`` (optional): regex to match against function names. Defaults to
+  ``.*``.
+* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
+* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 14]
```diff
@@ -331,6 +347,93 @@ The example below defines the following table access policy:
...
+  ``TABLES``, ``WINDOW``. When user define a rule for ``AGGREGATE``, ``SCALAR``
+  or ``WINDOW`` function, ``catalog`` and ``schema`` fields are disallowed.
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 15]
```diff
@@ -106,6 +113,15 @@ DELETE FROM                          all                delete
 UPDATE                               all                update
 ==================================== ========== ======= ==================== ===================================================
 
+Permissions required for executing functions and PTFs (Polymorphic Table Function):
+
+==================================== ========== ======================== ========================== ===============================================================================
+SQL command                          Catalog    Function permission      Function kind              Note
+==================================== ========== ======================== ========================== ===============================================================================
+SELECT function()                               execute, grant_execute*  aggregate, scalar, window  grant_execute is required when function is executed with view owner privileges
+SELECT FROM TABLE(table_function())  all        execute, grant_execute*  table                      grant_execute is required when PTF is executed with view owner privileges
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 3 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I used new `list-table`, please confirm if I did it right

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 16]
```diff
@@ -25,6 +25,13 @@ cluster in a single JSON file.
 Configuration
 -------------
 
+.. warning::
+
+    As of Trino 401 and later, access to PTFs (Polymorphic Table Function)
```
Comment (m57lyra) [inline docs/src/main/sphinx/security/file-system-access-control.rst:30]: Functions)

**Comment by huberty89** [comment]
@colebow @m57lyra Please take a look, I applied fixed based on all comments

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 17]
```diff
@@ -25,6 +25,13 @@ cluster in a single JSON file.
...
+    is allowed by default. In Trino versions 400 and earlier, access is denied
+    by default. To mitigate unwanted access, you must add a ``function``
+    :ref:`rule <system-file-function-rules>` to deny ``TABLE`` function type.
```
Comment (m57lyra) [inline docs/src/main/sphinx/security/file-system-access-control.rst:33]: deny the ``

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 18]
```diff
@@ -331,6 +359,94 @@ The example below defines the following table access policy:
       ]
     }
 
+.. _system-file-function-rules:
+
+Function rules
+^^^^^^^^^^^^^^
+
+These rules control the user's ability to execute SQL functions including PTFs.
```
Comment (m57lyra) [inline docs/src/main/sphinx/security/file-system-access-control.rst:367]: functions, including

**Review by m57lyra — APPROVED** [review]
Just a couple of minor fixes needed.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 16 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 17 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 18 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

**Review by colebow — APPROVED** [review]

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 19]
```diff
@@ -25,6 +25,13 @@ cluster in a single JSON file.
 Configuration
 -------------
 
+.. warning::
```
Comment (kokosing) [inline docs/src/main/sphinx/security/file-system-access-control.rst:28]: I think this should be part of the release notes. If we add warnings like that we may end up with messy docs.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 20]
```diff
@@ -106,6 +113,28 @@ DELETE FROM                          all                delete
...
+   * - ``SELECT function()``
+     -
+     - ``execute``, ``grant_execute*``
+     - ``aggregate``, ``scalar``, ``window``
+     - ``grant_execute`` is required when function is executed with view owner privileges.
+   * - ``SELECT FROM TABLE(table_function())``
+     - ``all``
+     - ``execute``, ``grant_execute*``
+     - ``table``
+     - ``grant_execute`` is required when PTF is executed with view owner privileges.
```
Comment (kokosing) [inline docs/src/main/sphinx/security/file-system-access-control.rst:136]: Is it well know what PTF means? I would prefer to use table function. Same in other places.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 21]
```diff
@@ -331,6 +359,94 @@ The example below defines the following table access policy:
...
+* ``schema`` (optional): ``regex`` to match against schema name. Defaults to ``.*``.
+* ``function`` (optional): ``regex`` to match against function names. Defaults to
+  ``.*``.
+* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
+* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
+  ``TABLES``, ``WINDOW``. When a user defines a rule for ``AGGREGATE``, ``SCALAR``
+  or ``WINDOW`` functions, the ``catalog`` and ``schema`` fields are disallowed.
+
+
+To restore previous behavior, which denied all PTFs from any catalog, use the following
```
Comment (kokosing) [inline docs/src/main/sphinx/security/file-system-access-control.rst:385]: What is "previous"? I would prefer to write to achieve X behavior, do y configuration, where you need to explain what X is.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 19 above)
Comment (m57lyra) [reply docs/src/main/sphinx/security/file-system-access-control.rst:28]: Agreed, and it should be in a breaking change section, as this will break expected security behaviors.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 22]
```diff
@@ -331,6 +359,94 @@ The example below defines the following table access policy:
...
+    }
+
+
+The following example allows the ``admin`` user to execute ``query`` PTFs from any
+catalog:
+
+.. note::
+
+    It is recommended to only allow admins to use the ``query`` polymorphic table
+    function.
```
Comment (kokosing) [inline docs/src/main/sphinx/security/file-system-access-control.rst:419]: Add "why"

Maybe it would be nice to have some best-practices somewhere. Where you would have config to grant access to PTF to admin, then create a (definer) view on top of it with admin as owner and granting access to view to regular users. Maybe we could explain that only somewhere.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 19 above)
Comment (m57lyra) [reply docs/src/main/sphinx/security/file-system-access-control.rst:28]: I was told though that there are no special sections, and release notes are generated from issues?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 20 above)
Comment (m57lyra) [reply docs/src/main/sphinx/security/file-system-access-control.rst:136]: are all table functions polymorphic? If not, we need to differentiate the two. Google style guide and numerous other documentation standards specify that you spell term out the first time it is used, with the acronym. Thereafter, you just use the acronym. When things have entered very common usage (like HTTP, unless it is actual documentation for that protocol), you don't have to spell it out.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 19 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:28]: Release notes are not generated. We write them .. but we do not have a dedicated breaking changes section. But this can be added .. probably was already .. @colebow ?

Also the warning is still appropriate .. maybe without the version info and past behavior though

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 20 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:136]: No .. PTF is not well known .. use "table functions" .. and @m57lyra .. not all table functions are polymorphic .. check out the video I created and shared recently

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 23]
```diff
@@ -117,10 +146,9 @@ items do not need to already exist as any potential permission makes the item
 visible. Specifically:
 
 * ``catalog``: Visible if user is the owner of any nested schema, has
-  permissions on any nested table, or has permissions to set session properties
-  in the catalog.
+  permissions on any nested table or PTF, or has permissions to set session properties in the catalog.
 * ``schema``: Visible if the user is the owner of the schema, or has permissions
-  on any nested table.
+  on any nested table or PTF.
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:151]: table function

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 24]
```diff
@@ -331,6 +359,94 @@ The example below defines the following table access policy:
...
+.. _system-file-function-rules:
+
+Function rules
+^^^^^^^^^^^^^^
+
+These rules control the user's ability to execute SQL functions, including PTFs.
+
+Each function rule is composed of the following fields:
+
+* ``user`` (optional): ``regex`` to match against user name. Defaults to ``.*``.
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:371]: what kind of regex? A SQL regex or a Java regex? Or something else. Also .. regex is not a keyword or so .. just say regular expression

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 21 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:385]: Reword this to say what the setup does.. and ignore the fact that this was the default in older releases.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 22 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:419]: I would even say .. remove that.. there is no reason for this to be best practice in general. Instead explain what the consequence is of this setup .. and earlier what happens by default

Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:419]: As in like .. if anyone is allowed to execute table functions.. then what?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 20 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:136]: Maybe make a link to https://trino.io/docs/current/functions/table.html in every occurrence of "PTF" ?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 22 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:419]: `query` table function which works as a query pass-though - it's bypass the whole access control so IMO it's quite important to at least mention about it

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 19 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:28]: I remove information about Trino versions

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 21 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Reworded

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 23 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Changed

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 25]
```diff
@@ -331,6 +359,94 @@ The example below defines the following table access policy:
       ]
     }
 
+.. _system-file-function-rules:
+
+Function rules
+^^^^^^^^^^^^^^
+
+These rules control the user's ability to execute SQL functions, including PTFs.
```
[the remark this answers is not in the document]
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I added links to specific function sections.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 20 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I linked every `table function` to https://trino.io/docs/current/functions/table.html

**Comment by huberty89** [comment]
@mosabua @kokosing updated, please take a look

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 19 above)
Comment (kokosing) [reply docs/src/main/sphinx/security/file-system-access-control.rst:28]: I think we still need this too in release notes.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 26]
```diff
@@ -25,6 +25,12 @@ cluster in a single JSON file.
 Configuration
 -------------
 
+.. warning::
+
+    Access to :doc:`table functions </functions/table>` is allowed by default.
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:30]: Isnt it allowed for all functions?

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 27]
```diff
@@ -331,6 +359,100 @@ The example below defines the following table access policy:
...
 
+.. _system-file-function-rules:
+
+Function rules
+^^^^^^^^^^^^^^
+
+These rules control the user's ability to execute SQL functions,
+such as :doc:`aggregate functions </functions/aggregate>`, scalar functions,
+:doc:`table functions </functions/table>` and :doc:`window functions </functions/window>`.
+
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:370]: Somewhere here we should add that by default all functions are allowed.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 28]
```diff
@@ -331,6 +359,100 @@ The example below defines the following table access policy:
...
+      ]
+    }
+
+
+The following example allows the ``admin`` user to execute ``query`` table
+function from any catalog:
+
+.. note::
+
+    It is recommended to only allow admins to use the ``query``
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:424]: Do not make this recommendation.. change to normal paragraph that says the following section disallows usage of the query table function. And maybe add a reason why that might make sense (query passthrough ignores access controls)

**Review by mosabua — CHANGES_REQUESTED** [review]
Let's update a bit more.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 26 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:30]: True, I wrote this warning because before access to table functions was always denied - there was no rules to control the access. Maybe change this into:
```
Access to all functions including table functions is allowed by default.
```
@mosabua WDYT?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 27 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:370]:
```
These rules control the user's ability to execute SQL all function kinds, such as...
```
?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 28 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Got it, rewritten.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 26 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:30]: yes

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 27 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst:370]: Sure .. but also say .. by default all functions are available.

Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:370]: When user does not define any `functions` rule then all functions are available. That's how the rest of rules works as well.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 22 above)
Comment (mosabua) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Yeah .. mention.. but not recommend as best practice.. its just important to understand that it is kind of a security hole otherwise ;-) .. so we need to explain that and not just have a recommendation without detailing the reasoning

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 29]
```diff
@@ -633,6 +751,24 @@ These rules apply to ``filter_environment`` and ``mask_environment``.
...
+* ``group`` (optional): regular expression to match against group names.
+  Defaults to ``.*``.
+* ``schema`` (optional): regular expression to match against schema name.
+  Defaults to ``.*``.
+* ``function`` (optional): regular expression to match against function names.
+  Defaults to ``.*``.
+* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
+* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
+  ``TABLES``, ``WINDOW``. When a user defines a rule for ``AGGREGATE``, ``SCALAR``
+  or ``WINDOW`` functions, the ``catalog`` and ``schema`` fields are disallowed.
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:770]: should we add why ... 

because these are globally defined functions 

or something like that

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 30]
```diff
@@ -331,6 +359,96 @@ The example below defines the following table access policy:
...
+        {
+          "privileges": [],
+          "function_kinds": [
+            "TABLE"
+          ]
+        }
+      ]
+    }
+
+It's a good practice to limit access to ``query`` table function
```
Comment (mosabua) [inline docs/src/main/sphinx/security/file-system-access-control.rst:417]: Fix wrapping of this paragraph .. but otherwise this is good now imho

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 29 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst:770]: Ok, added explanation why

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 30 above)
Comment (huberty89) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Fixed, I think

**Review by mosabua — APPROVED** [review]
Looks good now. @martint could you merge this before the release since the feature already merged with the 401 release.

**Review by kokosing — APPROVED** [review]
